sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -158,11 +169,6 @@ object TypeCoercion {
     findTightestCommonType(t1, t2)
       .orElse(findWiderTypeForDecimal(t1, t2))
       .orElse(stringPromotion(t1, t2))
-      .orElse((t1, t2) match {
-        case (ArrayType(et1, containsNull1), ArrayType(et2, containsNull2)) =>
-          findWiderTypeForTwo(et1, et2).map(ArrayType(_, containsNull1 || containsNull2))
-        case _ => None
-      })
```

Comment:
I think if we now check for ArrayTypes (including MapTypes) in `findTightestCommonType`, the match here becomes redundant. @mgaido91, @gczsjdy, does this thinking make sense to you both?

Reply:
it makes sense, but I'd love that in your implementation in `findTightestCommonType`, you would replicate this logic, ie. removing the `sameType` guard and using `findWiderTypeForTwo`, in order to allow casting an array of int to an array of long. What do you think?

Reply:
Sure, I think that's possible. In order to handle instances with and without string promotion, I think it may be necessary to add a boolean parameter, and then to handle the instances where the pointType/keyType and valueType may result in `None`, see https://github.com/apache/spark/pull/20010/files#diff-383a8cdd0a9c58cae68e0a79295520a3R105

To support the minor change in function signature for `findTightestCommonType`, I have to do some refactoring in the tests. Let me know if you think there's a cleaner implementation, but this seems to help localize like concerns into `findTightestCommonType`.

Reply:
Thanks for your illustration.

Reply:
My suggestion: we define a new function, like `findWiderTypeForArray`. This new function can provide 'findWider' functionality compared to the 'findTightestArray' part(which is basically the first commit in your PR). We won't break the original `findTightest` semantic in this way and the code is clean.

Reply:
I like your idea @gczsjdy!

Reply:
@gczsjdy I've taken a shot at implementing your suggestion with `findWiderTypeForTwoComplex`, which takes as an argument a `widerTypeFunc`, describing which widening behavior to apply to point types (should they permit promotion to string or not). 

Because `ArrayType` instances that would require widening the type could be nested in `StructType` and `MapType`, I think it's necessary to have more case matching than would be in `findWiderTypeForArray`, hence `findWiderTypeForTwoComplex`.

Reply:
I like your implementation, except for the concerns I raised in another thread. Have you tested locally?

Reply:
Yes, this PR should be ready for a Jenkins build.